// roadmapper/flux/stores/jamatree/JamaTreeActions.js
var JamaTreeConstants = require('./JamaTreeConstants'),
    JamaItemProxy = require('roadmapper/data/JamaItemProxy'),
    _ = require('lodash');

var JamaTreeActions = {
    loadProject: function(payload) {
        //load project id info
        JamaItemProxy.getProject(payload.projectId, _.bind(function(err, data) {
            this.dispatch(JamaTreeConstants.LOAD_PROJECT_FROM_ID_COMPLETE, data);
        }, this));
        
    },
    loadRootChildren: function(payload) {

        JamaItemProxy.getRootChildren(payload.projectId, _.bind(function(err, data) {
            
            this.dispatch(JamaTreeConstants.LOAD_CHILDREN_FROM_PARENT_ID, {
                projectId: payload.projectId,
                parent: null,
                items: data
            });

        }, this));
    },
    loadChildrenOfItem: function(payload) {
        //item id
        //project id.
        JamaItemProxy.getChildren(payload.itemId, _.bind(function(err, res) {
            this.dispatch(JamaTreeConstants.LOAD_CHILDREN_FROM_PARENT_ID, {
                projectId: payload.projectId,
                parent: payload.itemId,
                items: res.body.data
            });
        }, this));
    }
};

module.exports = JamaTreeActions;
