// roadmapper/flux/stores/roadmap/RoadmapStore.js
var Fluxxor = require('fluxxor'),
    RoadmapConstants = require('./RoadmapConstants'),
    Proxy = require('roadmapper/data/Proxy');

var RoadmapStore = Fluxxor.createStore({

    initialize: function() {
        this.bindActions(
            RoadmapConstants.ROADMAP_CREATE_ROADMAP, this.createRoadmap,
            RoadmapConstants.ROADMAP_SAVE_ROADMAP, this.saveRoadmap,
            RoadmapConstants.ROADMAP_DELETE_ROADMAP, this.deleteRoadmap,
            RoadmapConstants.ROADMAP_UPDATE_JAMA_ITEM, this.updateJamaItem
        )
    },

    createRoadmap: function() {

        this.emit('change');
    },
    saveRoadmap: function() {

        this.emit('change');
    },
    deleteRoadmap: function(payload) {

        this.emit('change');
    },
    updateJamaItem: function(payload) {
        this.emit('change');
    },
    getRoadmap: function(roadmapId) {
        return Proxy.getRoadmap(roadmapId);
    },
    getRoadmaps: function() {
        return Proxy.getRoadmaps();
    }
});

module.exports = RoadmapStore;
